File: 0208-implement-trie-prefix-tree/0208-implement-trie-prefix-tree.js

var Trie = function(value) {
    this.root = {}
};

/** 
 * @param {string} word
 * @return {void}
 */
Trie.prototype.insert = function(word) {
    let node = this.root;
    for (let i = 0; i < word.length; i++) {
        const char = word[i];
        if (!node[char]) node[char] = {};
        node = node[char];
    }
    node.isEnd = true;
};

/** 
 * @param {string} word
 * @return {boolean}
 */
Trie.prototype.search = function(word) {
    const node = this.searchNode(word);
    return node === null ? false : node.isEnd === true;
};

Trie.prototype.searchNode = function(word) {
    let node = this.root;
    for (let i = 0; i < word.length; i++) {
        const char = word[i];
        if (!node[char]) return null;
        node = node[char];
    }
    return node;
}

/** 
 * @param {string} prefix
 * @return {boolean}
 */
Trie.prototype.startsWith = function(prefix) {
    const node = this.searchNode(prefix);
    return node !== null;
};

/** 
 * Your Trie object will be instantiated and called as such:
 * var obj = new Trie()
 * obj.insert(word)
 * var param_2 = obj.search(word)
 * var param_3 = obj.startsWith(prefix)
 */

File: 0208-implement-trie-prefix-tree/0208-implement-trie-prefix-tree.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.end_char = False

class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        current = self.root
        for char in word:
            if char not in current.children:
                current.children[char] = TrieNode()
            current = current.children[char]
        current.end_char = True

    def search(self, word: str) -> bool:
        current = self.root
        for i in range(len(word)):
            if word[i] in current.children:
                current = current.children[word[i]]
            else:
                return False
        return current.end_char

    def startsWith(self, prefix: str) -> bool:
        current = self.root
        for i in range(len(prefix)):
            if prefix[i] in current.children:
                current = current.children[prefix[i]]
            else:
                return False
        return True


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)